File: 49 Group Anagram/49.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // Initialize an unordered_map to store sorted strings as keys and their anagrams as values
        unordered_map<string, vector<string>> hash;

        // Iterate through each string in the input vector
        for (const string& s : strs) {
            // Create a copy of the string to sort
            string sorted_s = s;
            // Sort the characters of the current string
            sort(sorted_s.begin(), sorted_s.end());

            // Add the original string to the vector of anagrams for this sorted string
            hash[sorted_s].push_back(s);
        }

        // Initialize a vector to store the final result
        vector<vector<string>> result;

        // Iterate through the hash map and add each group of anagrams to the result
        for (const auto& pair : hash) {
            result.push_back(pair.second);
        }

        // Return the result vector containing groups of anagrams
        return result;
    }
};

File: 49 Group Anagram/49.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # Initialize an empty dictionary to store sorted strings as keys and their anagrams as values
        hash = {}
        # Initialize an empty list to store the final result (not used in this implementation)
        result = []

        # Iterate through each string in the input list
        for i in strs:
            # Sort the characters of the current string and join them back into a string
            sorted_i = "".join(sorted(i))
            
            # If the sorted string is not in the hash dictionary
            if sorted_i not in hash:
                # Add it as a key with a list containing the original string as its value
                hash[sorted_i] = [i]
            else:
                # If the sorted string is already in the hash, append the original string to its list of anagrams
                hash[sorted_i].append(i)

        # Return the values of the hash dictionary as a list, which gives us groups of anagrams
        return list(hash.values())